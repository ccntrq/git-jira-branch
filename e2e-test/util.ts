import {type SpawnSyncReturns, execSync, spawnSync} from 'node:child_process';
import {rmSync} from 'node:fs';
import {mkdtemp} from 'node:fs/promises';
import {tmpdir} from 'node:os';
import {join} from 'node:path';
import {Brand} from 'effect';

export type Directory = string & Brand.Brand<'Directory'>;
export const Directory = Brand.nominal<Directory>();

export const setupTmpDir = async (): Promise<
  readonly [Directory, () => void]
> => {
  const tmpDir = await mkdtemp(join(tmpdir(), 'git-jira-branch-e2e-'));
  // biome-ignore lint/nursery/noConsole: log statement okay in this test
  console.debug(`Setting up e2e env in ${tmpDir}`);

  execSync('git init', {cwd: tmpDir});

  execSync('git config user.email "lester-tester@example.com"', {
    cwd: tmpDir,
  });
  execSync('git config user.name "Lester Tester"', {cwd: tmpDir});
  execSync('git commit -m "init" --allow-empty', {cwd: tmpDir});

  const cleanup = () => {
    // biome-ignore lint/nursery/noConsole: log statement okay in test
    console.debug(`Cleaning up e2e env in ${tmpDir}`);
    rmSync(tmpDir, {force: true, recursive: true});
  };

  return [Directory(tmpDir), cleanup] as const;
};

export const runApp = (dir: Directory, ...args: Array<string>): string => {
  const cmd = `git-jira-branch ${args.join(' ')}`;
  return execSync(cmd, {cwd: dir}).toString();
};

export const spawnApp = (
  dir: Directory,
  ...args: Array<string>
): SpawnSyncReturns<Buffer> => {
  const cmd = `git-jira-branch ${args.join(' ')}`;
  return spawnSync(cmd, {
    cwd: dir,
    shell: true,
  });
};

export const currentBranch = (dir: Directory): string =>
  execSync('git branch --show-current', {cwd: dir}) //
    .toString()
    .trim();

export const createBranch = (dir: Directory, name: string): string =>
  execSync(`git checkout -b ${name}`, {cwd: dir}) //
    .toString()
    .trim();

export const switchBranch = (dir: Directory, name: string): string =>
  execSync(`git switch ${name}`, {cwd: dir}) //
    .toString()
    .trim();

export const createCommit = (dir: Directory, message: string): string =>
  execSync(`git commit -m "${message}" --allow-empty`, {cwd: dir}) //
    .toString()
    .trim();
