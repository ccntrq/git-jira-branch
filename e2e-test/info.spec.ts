import {beforeEach, describe, expect, it} from 'vitest';
import {type Directory, createBranch, runApp, setupTmpDir} from './util';

describe('git-jira-branch info', () => {
  let tmpDir: Directory;
  const infoCommand = (...args: Array<string>) =>
    runApp(tmpDir, 'info', ...args);

  beforeEach(async () => {
    const [dir, cleanup] = await setupTmpDir();
    tmpDir = dir;
    return cleanup;
  });
  it('starts app and outputs help', async () => {
    const res = await infoCommand('--help');
    expect(res).toMatch(/git-jira-branch/);
  });

  it('prints ticket info for given ticket', async () => {
    const result = await infoCommand('1');
    expect(result).toMatchSnapshot();
  });

  it('info subcommand prints ticket info for current branch', async () => {
    createBranch(tmpDir, 'feat/GCJB-1-e2e-test-ticket-with-a-fancy-summary');
    const result = await infoCommand();
    expect(result).toMatchSnapshot();
  });
});
