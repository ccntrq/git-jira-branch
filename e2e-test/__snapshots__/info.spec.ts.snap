// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`git-jira-branch info > info subcommand prints ticket info for current branch 1`] = `
"[0;4m[0;1;4mGCJB-1[0;4m - E2E Test ticket with a fancy summary [0m
[0;4m[0;1;4mTask[0;4m | [0;1;4mStatus:[0;4m To Do | [0;1;4mCreator:[0;4m Alexander Pankoff | [0;1;4mAssignee:[0;4m Alexander Pankoff[0m

*!!\\!IMPORTANT\\!!!*  - Do not modify this ticket without updating
{{git-jira-branch}} E2E tests aswell!


The description of this E2E Test ticket is even more fancy

In the description, there exist certain lines of considerable length that,
despite their extensive character count, do not contribute any meaningful or
valuable information whatsoever.

* a list item
** a sublist item
** {{inline code}}
** _Italics_
** +Underlined+
** -Strikethrough-

||*Table Row 1*||*Table Row 2*||
|Cell 1|Cell 2|
|Cell 3|Cell 4|

{noformat}const sourceCode: string = "some ts source code";{noformat}
"
`;

exports[`git-jira-branch info > prints ticket info for given ticket 1`] = `
"[0;4m[0;1;4mGCJB-1[0;4m - E2E Test ticket with a fancy summary [0m
[0;4m[0;1;4mTask[0;4m | [0;1;4mStatus:[0;4m To Do | [0;1;4mCreator:[0;4m Alexander Pankoff | [0;1;4mAssignee:[0;4m Alexander Pankoff[0m

*!!\\!IMPORTANT\\!!!*  - Do not modify this ticket without updating
{{git-jira-branch}} E2E tests aswell!


The description of this E2E Test ticket is even more fancy

In the description, there exist certain lines of considerable length that,
despite their extensive character count, do not contribute any meaningful or
valuable information whatsoever.

* a list item
** a sublist item
** {{inline code}}
** _Italics_
** +Underlined+
** -Strikethrough-

||*Table Row 1*||*Table Row 2*||
|Cell 1|Cell 2|
|Cell 3|Cell 4|

{noformat}const sourceCode: string = "some ts source code";{noformat}
"
`;
